src/vs/workbench/contrib/codeEditor/browser/saveParticipants.ts

```diff
@@ -318,11 +317,29 @@ class CodeActionOnSaveParticipant implements ITextFileSaveParticipant {
 		const excludedActions = Array.isArray(setting)
 			? []
 			: Object.keys(setting)
-				.filter(x => setting[x] === false)
+				.filter(x => setting[x] === 'never')
 				.map(x => new CodeActionKind(x));
 
+		const explicitActions = Array.isArray(setting)
+			? []
+			: Object.keys(setting)
+				.filter(x => setting[x] === 'explicit')
+				.map(x => new CodeActionKind(x));
+
+
 		progress.report({ message: localize('codeaction', "Quick Fixes") });
-		await this.applyOnSaveActions(textEditorModel, codeActionsOnSave, excludedActions, progress, token);
+
+		let filteredSaveList = codeActionsOnSave.filter(item => !excludedActions.some(matchKind => this.areCodeActionKindsEqual(item, matchKind)));
+
+		if (env.reason !== SaveReason.EXPLICIT) {
```

Comment:
I think this code can be simplified. Maybe something like:

```
let actionsToRun =  ... .filter(x => setting[x] === 'explicit' && env.reason == SaveReason.EXPLICIT || setting[x] === 'always') 
```

That way you do all the filtering in a single place